crates/hotshot-stake-table/src/vec_based.rs

```diff
@@ -0,0 +1,436 @@
+//! A vector based stake table implementation. The commitment is the rescue hash of the list of (key, amount) pairs;
+
+use crate::utils::{u256_to_field, ToFields};
+use ark_std::{collections::HashMap, hash::Hash, rand::SeedableRng};
+use digest::crypto_common::rand_core::CryptoRngCore;
+use ethereum_types::{U256, U512};
+use hotshot_types::traits::stake_table::{SnapshotVersion, StakeTableError, StakeTableScheme};
+use jf_primitives::rescue::{sponge::RescueCRHF, RescueParameter};
+use serde::{Deserialize, Serialize};
+
+pub mod config;
+
+/// Locally maintained stake table, generic over public key type `K`.
+/// Whose commitment is a rescue hash of all key-value pairs over field `F`.
+/// NOTE: the commitment is only available for the finalized versions, and is
+/// computed only once when it's finalized.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct StakeTable<K: Eq + Hash + Clone + ToFields<F>, F: RescueParameter> {
+    /// The most up-to-date stake table, where the incoming transactions shall be performed on.
+    head: Vec<(K, U256)>,
+    /// The snapshot of stake table at the beginning of the current epoch
+    epoch_start: Vec<(K, U256)>,
+    /// The stake table used for leader election.
+    last_epoch_start: Vec<(K, U256)>,
+
+    /// Total stakes for different versions
+    head_total_stake: U256,
+    epoch_start_total_stake: U256,
+    last_epoch_start_total_stake: U256,
+
+    /// Commitment for finalized versions
+    epoch_start_comm: F,
+    last_epoch_start_comm: F,
+
+    /// The mapping from public keys to their location in the Merkle tree.
+    #[serde(skip)]
+    mapping: HashMap<K, usize>,
+}
+
+impl<K, F> StakeTableScheme for StakeTable<K, F>
+where
+    K: Eq + Hash + Clone + ToFields<F>,
+    F: RescueParameter,
+{
+    type Key = K;
+    type Amount = U256;
+    type Commitment = F;
+    type LookupProof = ();
+    // TODO(Chengyu): Can we make it references?
+    type IntoIter = <Vec<(K, U256)> as ark_std::iter::IntoIterator>::IntoIter;
+    // type IntoIter = ark_std::slice::Iter<'a, &'a (K, U256)>;
+
+    fn register(
+        &mut self,
+        new_key: &Self::Key,
+        amount: Self::Amount,
+    ) -> Result<(), StakeTableError> {
+        match self.mapping.get(new_key) {
+            Some(_) => Err(StakeTableError::ExistingKey),
+            None => {
+                let pos = self.mapping.len();
+                self.head.push((new_key.clone(), amount));
+                self.head_total_stake += amount;
+                self.mapping.insert(new_key.clone(), pos);
+                Ok(())
+            }
+        }
+    }
+
+    fn deregister(&mut self, existing_key: &Self::Key) -> Result<(), StakeTableError> {
+        match self.mapping.get(existing_key) {
+            Some(pos) => {
+                self.head_total_stake -= self.head[*pos].1;
+                self.head[*pos].1 = U256::zero();
```

Comment:
should we _remove_ or set it to zero here? 
the former clean up the memory usage and will affect updated stake table commitment.

Reply:
This is a debatable question. I vaguely remember that @jbearer brought this idea up before. We may settle for it right now, or we may have to implement some garbage collection mechanism.